### ISSUE TITLE
[Bug] EAGLE greedy verify lacks TP broadcast → ranks diverge on accepted tokens → collective deadlock (tp>1)

### ISSUE THREAD

**Issue by zhoaa** [sgl-project/sglang#31071]
### Describe the bug

In EAGLE / spec-v2 verify (`python/sglang/srt/speculative/eagle_utils.py`), the **greedy** branch computes the accepted-token set from a **per-rank local `torch.argmax(next_token_logits)`** and does **not** broadcast the result across tensor-parallel ranks. The sibling **sampling** branch *does* broadcast `predict` / `accept_index` / `num_correct_drafts` from rank 0, with a comment explaining ranks can disagree due to FP non-determinism.

- greedy branch (no broadcast): https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/speculative/eagle_utils.py#L662-L675
- sampling branch (has broadcast): https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/speculative/eagle_utils.py#L758-L770

The greedy-branch condition includes `_is_hip` (and `_is_cpu`/`_is_npu`/`_is_xpu`), so those backends **always** take the no-broadcast path (even for temperature > 0 requests). When the two TP ranks' `next_token_logits` differ slightly — FP non-determinism, which the sampling branch's own comment acknowledges ("different GPUs may produce slightly different … causing different sampled tokens") — and a logit is **near-tied**, `argmax` picks a **different token on each rank** → the ranks **accept a different number of draft tokens** → committed `seq_lens` / batch shapes diverge across ranks → **the next tensor-parallel collective deadlocks.**

The failure is a **silent, permanent hang**: both scheduler ranks stuck at

```
Event.synchronize     (torch/cuda/streams.py)
resolve_seq_lens_cpu  (srt/managers/overlap_utils.py)
run_batch             (srt/managers/scheduler.py)
```

with no exception/error, `num_running_reqs` frozen and generation throughput → 0, while the HTTP frontend still returns 200 on `/health` (so a cheap liveness probe does not detect it). The scheduler watchdog eventually kills the process.

### Root cause

The greedy verify path assumes *identical logits ⇒ identical argmax across ranks ⇒ no broadcast needed*. That holds when the all-reduce delivers bit-identical results to every rank (e.g. deterministic NCCL) but breaks when per-rank logits differ (non-deterministic reduction — e.g. AMD `--enable-aiter-allreduce-fusion`). The sampling branch guards against this with a rank-0 broadcast; the greedy branch does not.

### Reproduction

Mechanism isolation (deterministic): on a `tp>1` EAGLE deployment, force one rank's post-argmax `target_predict` to differ from the other rank's for a single verify step (simulating a near-tie that argmax resolves differently per rank). The ranks then accept a different number of tokens and the **next TP collective deadlocks** — both ranks wedge in `resolve_seq_lens_cpu → Event.synchronize`, `num_running_reqs` frozen, gen tok/s → 0, `/health` still 200. Reverting to a single, cross-rank-consistent accept decision (i.e. adding the broadcast) prevents the hang.

Observed in production on **sglang v0.5.12.post1** (ROCm, MI350X/gfx950), **tp=2**, EAGLE (`--speculative-num-steps 3`), `--attention-backend triton`, `--enable-aiter-allreduce-fusion`. The code path is also present on `main` (@ `e2728ac`).

### Suggested fix

Mirror the sampling branch's broadcast into the greedy branch:

```python
if tp_group.world_size > 1:
    tp_group.broadcast(predict, src=0)
    tp_group.broadcast(accept_index, src=0)
    tp_group.broadcast(num_correct_drafts, src=0)
```

so the greedy accept decision is TP-consistent regardless of per-rank logit differences.

### Possibly related

This looks like the same root cause as #24440 (`[Bug][AMD] Accuracy issue with EAGLE speculative decoding … Qwen3.5`) — per-rank verify divergence — surfacing as a **wrong-output accuracy** bug in the common case, and as this **collective deadlock** in the rarer case where the divergent accept *count* desyncs a collective.

### Environment

- sglang v0.5.12.post1 (ROCm 7.2, MI35x image); path also present on `main`.
- AMD MI350X (gfx950), tp=2.
- Speculative decoding: EAGLE, `--speculative-num-steps 3`, `--speculative-eagle-topk 1`.
- `--attention-backend triton`, `--enable-aiter-allreduce-fusion`.

**Comment by Allyyi** [comment]
Thanks for detailed analysis. I'd like to reproduce and work on it

**Comment by Allyyi** [comment]
Hi @zhoaa , I can't reproduce it because it's really hard to find available cloud AMD gpu rental. But I do agree your fix for adding broadcast to greedy branch. On top of this, I suggest adding broadcast to a unified place after the if block at https://github.com/sgl-project/sglang/blob/5cbea10e2fcd269809c26237c175a470719fc9fe/python/sglang/srt/speculative/eagle_utils.py#L762, so the `predict`, `accept_index`, and `num_correct_drafts` will reach to consistence after verified decision is finalized before they get consumed by worker, covering the greedy path, the sampling path, and any later changes before return.
